File: perceptron.py
import numpy as np


class Perceptron():
    def __init__(self, w=[], b=0.0, layer=None):
        self.layer = layer
        self.w = np.array(w)
        self.b = b

    def __multi_layer_call(self, x1: int or float, x2: int or float, f_list):
        """再帰処理用"""
        try:
            # 再帰的にパーセプトロンを呼び出し
            return f_list[0].call(
                    self.__multi_layer_call(x1, x2, f_list=f_list[1][0]),
                    self.__multi_layer_call(x1, x2, f_list=f_list[1][1])
                )
        except TypeError:
            # 末尾
            return f_list.call(x1, x2)

    def call(self, x1, x2) -> bool:
        if not self.layer:
            # Single layer
            x = np.array([x1, x2])

            if np.sum(self.w * x) + self.b <= 0:
                return False
            else:
                # 発火
                return True

        else:
            # Multi layer
            return self.__multi_layer_call(x1, x2, self.layer)


if __name__ == "__main__":
    AND = Perceptron(w=[0.5, 0.5], b=-0.7)
    NAND = Perceptron(w=[-0.5, -0.5], b=0.7)
    OR = Perceptron(w=[1, 1], b=-0.5)
    XOR = Perceptron(layer=[AND, [NAND, OR]])

    xor_table = [
        XOR.call(0, 0),
        XOR.call(0, 1),
        XOR.call(1, 0),
        XOR.call(1, 1),
    ]
    print(xor_table)

File: nim/2/perceptron.nim
import ../utils
import sequtils

type Perceptron = ref object
    layer: seq[seq[Perceptron]]
    w: array[2, float]
    b: float


proc call(p: Perceptron, x1: bool, x2: bool): bool =
    if p.layer.len == 0:
        # single layer
        var sum: float = 0.0
        let x = [float(x1), float(x2)]
        for item in zip(p.w, x):
            sum += item[0] * item[1]

        if sum+p.b > 0:
            result = true
        else:
            result = false
    else:
        result = p.layer[0][0].call(p.layer[1][0].call(x1, x2), p.layer[1][1].call(x1, x2))

const table = @[@[false, false],
                @[false, true],
                @[true, false],
                @[true, true]]

let AND: Perceptron = Perceptron(w: [0.5, 0.5], b: -0.7)
let NAND: Perceptron = Perceptron(w: [-0.5, -0.5], b: 0.7)
let OR: Perceptron = Perceptron(w: [1.0, 1.0], b: -0.5)
let XOR: Perceptron = Perceptron(layer: @[@[AND], @[NAND, OR]])
let and_table = table.map do (x: seq[bool]) -> bool: AND.call(x[0], x[1])
let nand_table = table.map do (x: seq[bool]) -> bool: NAND.call(x[0], x[1])
let or_table = table.map do (x: seq[bool]) -> bool: OR.call(x[0], x[1])
let xor_table = table.map do (x: seq[bool]) -> bool: XOR.call(x[0], x[1])
echo and_table
echo nand_table
echo or_table
echo xor_table
